File: php/src/Address.php
<?php
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
declare(strict_types=1);

namespace Zoobc\Zbc;

/** Account types, the ZBC_/ZNK_/ZBS_ text form and every recipient form of spec/addresses.md. */
final class Address
{
    public const ZOOBC = 0, BITCOIN = 1, EMPTY = 2, ESTONIA_EID = 3, ETHEREUM = 4, BITCOIN_P2PKH = 5, BITCOIN_P2SH = 6, BITCOIN_P2WPKH = 7;
    public const BITCOIN_P2WSH = 8, BITCOIN_TAPROOT = 9, DATASET = 10, SOLANA = 11, POLKADOT = 12, CARDANO = 13, RIPPLE = 14, TRON = 15, TEZOS = 16;
    private const NAMES = [0 => 'ZooBC', 1 => 'Bitcoin', 3 => 'Estonia eID', 4 => 'Ethereum', 5 => 'Bitcoin P2PKH', 6 => 'Bitcoin P2SH', 7 => 'Bitcoin P2WPKH',
        8 => 'Bitcoin P2WSH', 9 => 'Bitcoin Taproot', 10 => 'DataSet', 11 => 'Solana', 12 => 'Polkadot', 13 => 'Cardano', 14 => 'Ripple', 15 => 'Tron', 16 => 'Tezos'];
    private const CHAINS = ['zbc' => 'zbc', 'zoobc' => 'zbc', 'btc' => 'btc', 'bitcoin' => 'btc', 'eth' => 'eth', 'ethereum' => 'eth', 'evm' => 'eth',
        'sol' => 'sol', 'solana' => 'sol', 'dot' => 'dot', 'polkadot' => 'dot', 'substrate' => 'dot', 'ada' => 'ada', 'cardano' => 'ada',
        'xrp' => 'xrp', 'ripple' => 'xrp', 'trx' => 'trx', 'tron' => 'trx', 'xtz' => 'xtz', 'tezos' => 'xtz', 'zbs' => 'zbs', 'dataset' => 'zbs'];

    public static function typeName(int $t): string { return self::NAMES[$t] ?? "type $t"; }
    public static function typed(int $t, string $payload): string { return Encoding::le32($t) . $payload; }

    /** PREFIX_ + base32(payload || SHA3-256(payload || prefix)[0..2]) in seven groups of eight. */
    public static function encode(string $payload, string $prefix = 'ZBC'): string
    {
        if (strlen($payload) !== 32 || strlen($prefix) !== 3) { throw new \InvalidArgumentException('address payload must be 32 bytes and the prefix 3 characters'); }
        $s = Encoding::base32Encode($payload . substr(Encoding::sha3($payload, $prefix), 0, 3));
        $out = $prefix;
        for ($i = 0; $i < 7; $i++) { $out .= '_' . substr($s, 8 * $i, 8); }
        return $out;
    }

    /** [upper-case prefix, 32-byte payload] of PREFIX_... (separators _ or -, any case), or null. */
    /** The 59 significant characters of a ZooBC address: separators (_ -) and whitespace dropped, upper case (addresses.md 2). */
    public static function significant(string $text): string
    {
        return strtoupper((string) preg_replace('/[-_\s]/', '', $text));
    }

    /** [upper-case prefix, 32-byte payload] of a ZooBC address in any spelling, or null. */
    public static function decode(string $text): ?array
    {
        $norm = self::significant($text);
        if (strlen($norm) < 3) { return null; }
        $prefix = substr($norm, 0, 3);
        $body = substr($norm, 3);
        if (strlen($body) !== 56) { return null; }
        $raw = Encoding::base32Decode($body);
        if ($raw === null || strlen($raw) !== 35) { return null; }
        $payload = substr($raw, 0, 32);
        return substr(Encoding::sha3($payload, $prefix), 0, 3) === substr($raw, 32) ? [$prefix, $payload] : null;
    }

    /** Shape only: PREFIX then a separator, or the bare form: 59 significant characters, ZBC/ZBS prefix, base32 body. */
    private static function looksZbc(string $a): bool
    {
        if (strlen($a) > 4 && ($a[3] === '_' || $a[3] === '-')) { return true; }
        $n = self::significant($a);
        return strlen($n) === 59 && (str_starts_with($n, 'ZBC') || str_starts_with($n, 'ZBS')) && preg_match('/^[A-Z2-7]{56}$/', substr($n, 3)) === 1;
    }

    private static function zbcForm(string $a): ParsedAddress
    {
        $d = self::decode($a);
        if ($d === null) { throw new \InvalidArgumentException('invalid ZooBC address checksum'); }
        return new ParsedAddress($d[0] === 'ZBS' ? self::DATASET : self::ZOOBC, $d[1], $a);
    }

    private static function hinted(string $a, string $hint): ParsedAddress
    {
        switch ($hint) {
            case 'eth':
                $h = preg_replace('/^0[xX]/', '', $a);
                if (!Encoding::isHex($h, 40)) { throw new \InvalidArgumentException('not a 20-byte Ethereum address'); }
                return new ParsedAddress(self::ETHEREUM, hex2bin($h), $a);
            case 'sol':
                $d = Encoding::base58Decode($a);
                if ($d === null || strlen($d) !== 32) { throw new \InvalidArgumentException('not a 32-byte Solana address'); }
                return new ParsedAddress(self::SOLANA, $d, $a);
            case 'dot':
                $ss = Encoding::ss58Decode($a);
                if ($ss === null) { throw new \InvalidArgumentException('not a valid SS58 address'); }
                return new ParsedAddress(self::POLKADOT, $ss[1], $a);
            case 'zbc': case 'zbs':
                return self::zbcForm($a);
        }
        return self::auto($a);
    }

    private static function auto(string $a): ParsedAddress
    {
        if (strlen($a) === 42 && preg_match('/^0[xX]/', $a) && Encoding::isHex(substr($a, 2), 40)) { return new ParsedAddress(self::ETHEREUM, hex2bin(substr($a, 2)), $a); }
        if (self::looksZbc($a)) { return self::zbcForm($a); }
        $low5 = strtolower(substr($a, 0, 5));
        if (str_starts_with($low5, 'bc1') || str_starts_with($low5, 'tb1') || str_starts_with($low5, 'bcrt1')) {
            $d = Encoding::segwitDecode($a);
            if ($d === null) { throw new \InvalidArgumentException('invalid Bitcoin bech32 address'); }
            [, $version, $prog] = $d;
            if ($version === 0 && strlen($prog) === 20) { return new ParsedAddress(self::BITCOIN_P2WPKH, $prog, $a); }
            if ($version === 0 && strlen($prog) === 32) { return new ParsedAddress(self::BITCOIN_P2WSH, $prog, $a); }
            if ($version === 1 && strlen($prog) === 32) { return new ParsedAddress(self::BITCOIN_TAPROOT, $prog, $a); }
            throw new \InvalidArgumentException('unsupported Bitcoin witness program');
        }
        if (($a[0] === '1' || $a[0] === '3') && strlen($a) >= 26 && strlen($a) <= 35) {
            $raw = Encoding::base58Decode($a);
            if ($raw !== null && strlen($raw) === 25) {
                $body = Encoding::base58CheckDecode($a);
                if ($body !== null && strlen($body) === 21) {
                    if ($body[0] === "\x00") { return new ParsedAddress(self::BITCOIN_P2PKH, substr($body, 1), $a); }
                    if ($body[0] === "\x05") { return new ParsedAddress(self::BITCOIN_P2SH, substr($body, 1), $a); }
                }
            }
        }
        if (strlen($a) > 5 && strtolower(substr($a, 0, 5)) === 'addr1') {
            $d = Encoding::bech32DecodePlain($a);
            if ($d !== null && $d[0] === 'addr' && strlen($d[1]) === 29 && $d[1][0] === "\x61") { return new ParsedAddress(self::CARDANO, substr($d[1], 1), $a); }
            throw new \InvalidArgumentException('invalid Cardano address (expected a mainnet enterprise addr1… address)');
        }
        if ($a[0] === 'T' && strlen($a) === 34) {
            $body = Encoding::base58CheckDecode($a);
            if ($body !== null && strlen($body) === 21 && $body[0] === "\x41") { return new ParsedAddress(self::TRON, substr($body, 1), $a); }
            throw new \InvalidArgumentException('invalid Tron address');
        }
        if ($a[0] === 'r' && strlen($a) >= 25 && strlen($a) <= 35) {
            $body = Encoding::base58CheckDecode($a, Encoding::RIPPLE_ALPHABET);
            if ($body !== null && strlen($body) === 21 && $body[0] === "\x00") { return new ParsedAddress(self::RIPPLE, substr($body, 1), $a); }
            throw new \InvalidArgumentException('invalid Ripple address');
        }
        if (str_starts_with($a, 'tz1')) {
            $body = Encoding::base58CheckDecode($a);
            if ($body !== null && strlen($body) === 23 && substr($body, 0, 3) === "\x06\xa1\x9f") { return new ParsedAddress(self::TEZOS, substr($body, 3), $a); }
            throw new \InvalidArgumentException('invalid Tezos address');
        }
        $ss = Encoding::ss58Decode($a);
        if ($ss !== null && strlen($ss[1]) === 32) { return new ParsedAddress(self::POLKADOT, $ss[1], $a); }
        if (strlen($a) >= 32 && strlen($a) <= 44) {
            $d = Encoding::base58Decode($a);
            if ($d !== null && strlen($d) === 32) { return new ParsedAddress(self::SOLANA, $d, $a); }
        }
        if (Encoding::isHex($a, 64)) { $key = hex2bin($a); return new ParsedAddress(self::ZOOBC, $key, self::encode($key, 'ZBC')); }
        throw new \InvalidArgumentException('unrecognised address. Supported: ZooBC (ZBC_/ZBS_), Bitcoin, Ethereum, Solana, Polkadot, Cardano, Ripple, Tron, Tezos');
    }

    /** Read a recipient in the order of spec/addresses.md section 3; $chain forces one reading (--chain). Throws InvalidArgumentException. */
    public static function parse(string $text, string $chain = ''): ParsedAddress
    {
        $a = trim($text);
        if ($a === '') { throw new \InvalidArgumentException('empty address'); }
        if ($chain === '') { return self::auto($a); }
        $hint = self::CHAINS[strtolower($chain)] ?? throw new \InvalidArgumentException("unknown chain $chain");
        try { return self::hinted($a, $hint); } catch (\InvalidArgumentException $e) {
            $up = strtoupper($a); $low = strtolower($a);
            $plain = (strlen($a) === 42 && preg_match('/^0[xX]/', $a)) || str_starts_with($up, 'ZBC') || str_starts_with($up, 'ZNK') || str_starts_with($up, 'ZBS')
                || $a[0] === '1' || $a[0] === '3' || str_starts_with($low, 'bc1') || str_starts_with($low, 'tb1') || str_starts_with($low, 'bcrt1') || strlen($a) === 64;
            if (!$plain) { throw $e; }
            return self::auto($a);
        }
    }

    /** A registry key parameter: 64 hex (optionally 0x) or a ZNK_/ZBG_/ZBR_/ZBC_ text address; 32 bytes. */
    public static function parseKey32(string $text): string
    {
        if (strlen($text) === 66 && $text[3] === '_') {
            $d = self::decode($text);
            if ($d === null) { throw new \InvalidArgumentException('invalid address checksum'); }
            return $d[1];
        }
        $h = preg_replace('/^0[xX]/', '', $text);
        if (!Encoding::isHex($h, 64)) { throw new \InvalidArgumentException('key must be a 64-hex string or a ZNK_/ZBG_/ZBR_ address'); }
        return hex2bin($h);
    }
}

File: swift/Sources/ZBC/Address.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci

/// Account types (spec/addresses.md section 1).
public enum AccountType {
    public static let zoobc: Int32 = 0, bitcoin: Int32 = 1, empty: Int32 = 2, estoniaEID: Int32 = 3, ethereum: Int32 = 4
    public static let bitcoinP2PKH: Int32 = 5, bitcoinP2SH: Int32 = 6, bitcoinP2WPKH: Int32 = 7, bitcoinP2WSH: Int32 = 8, bitcoinTaproot: Int32 = 9
    public static let dataset: Int32 = 10, solana: Int32 = 11, polkadot: Int32 = 12, cardano: Int32 = 13, ripple: Int32 = 14, tron: Int32 = 15, tezos: Int32 = 16

    public static func name(_ t: Int32) -> String {
        switch t {
        case 0: return "ZooBC"; case 1: return "Bitcoin"; case 3: return "Estonia eID"; case 4: return "Ethereum"; case 5: return "Bitcoin P2PKH"
        case 6: return "Bitcoin P2SH"; case 7: return "Bitcoin P2WPKH"; case 8: return "Bitcoin P2WSH"; case 9: return "Bitcoin Taproot"; case 10: return "DataSet"
        case 11: return "Solana"; case 12: return "Polkadot"; case 13: return "Cardano"; case 14: return "Ripple"; case 15: return "Tron"; case 16: return "Tezos"
        default: return "type \(t)"
        }
    }
}

/// A recipient as the envelope carries it.
public struct ParsedAddress: Equatable {
    public let type: Int32
    public let payload: [UInt8]
    public let display: String
    /// 4-byte type LE then the payload (36 bytes for a ZooBC account).
    public var bytes: [UInt8] { Address.typed(type, payload) }
    public var typeName: String { AccountType.name(type) }
}

/// The ZBC_/ZNK_/ZBS_ text form and every recipient form of spec/addresses.md.
public enum Address {
    public static func typed(_ t: Int32, _ payload: [UInt8]) -> [UInt8] { LE.i32(t) + payload }

    /// PREFIX_ + base32(payload || SHA3-256(payload || prefix)[0..2]) in seven groups of eight.
    public static func encode(_ payload: [UInt8], prefix: String = "ZBC") -> String? {
        guard payload.count == 32, prefix.utf8.count == 3 else { return nil }
        let check = Array(SHA3.hash256(payload, Array(prefix.utf8))[0..<3])
        let s = Array(Enc.base32Encode(payload + check).utf8)
        var out = prefix
        for i in 0..<7 { out += "_" + String(decoding: s[(8 * i)..<(8 * i + 8)], as: UTF8.self) }
        return out
    }

    /// The 59 significant characters of a ZooBC address: separators (_ -) and whitespace dropped, upper case (addresses.md 2).
    public static func significant(_ text: String) -> [UInt8] {
        return Array(text.uppercased().utf8).filter { c in
            !(c == UInt8(ascii: "_") || c == UInt8(ascii: "-") || c == 0x20 || (c >= 0x09 && c <= 0x0d))
        }
    }

    /// (upper-case prefix, 32-byte payload) of a ZooBC address in any spelling, or nil.
    public static func decode(_ text: String) -> (String, [UInt8])? {
        let norm = significant(text)
        guard norm.count >= 3 else { return nil }
        let prefix = String(decoding: norm[0..<3], as: UTF8.self)
        let body = Array(norm[3...])
        guard body.count == 56, let raw = Enc.base32Decode(String(decoding: body, as: UTF8.self)), raw.count == 35 else { return nil }
        let payload = Array(raw[0..<32])
        let check = Array(SHA3.hash256(payload, Array(prefix.utf8))[0..<3])
        return check == Array(raw[32..<35]) ? (prefix, payload) : nil
    }

    private static let chains: [String: String] = ["zbc": "zbc", "zoobc": "zbc", "btc": "btc", "bitcoin": "btc", "eth": "eth", "ethereum": "eth", "evm": "eth",
        "sol": "sol", "solana": "sol", "dot": "dot", "polkadot": "dot", "substrate": "dot", "ada": "ada", "cardano": "ada", "xrp": "xrp", "ripple": "xrp",
        "trx": "trx", "tron": "trx", "xtz": "xtz", "tezos": "xtz", "zbs": "zbs", "dataset": "zbs"]

    public struct Invalid: Error { public let message: String }

    private static let b32Alphabet = Set("ABCDEFGHIJKLMNOPQRSTUVWXYZ234567".utf8)

    /// Shape only: PREFIX then a separator, or the bare form: 59 significant characters, ZBC/ZBS prefix, base32 body.
    private static func looksZbc(_ a: String) -> Bool {
        let u = Array(a.utf8)
        if u.count > 4 && (u[3] == UInt8(ascii: "_") || u[3] == UInt8(ascii: "-")) { return true }
        let n = significant(a)
        guard n.count == 59 else { return false }
        let p = String(decoding: n[0..<3], as: UTF8.self)
        return (p == "ZBC" || p == "ZBS") && n[3...].allSatisfy { b32Alphabet.contains($0) }
    }

    private static func zbcForm(_ a: String) throws -> ParsedAddress {
        guard let d = decode(a) else { throw Invalid(message: "invalid ZooBC address checksum") }
        let (prefix, payload) = d
        return ParsedAddress(type: prefix == "ZBS" ? AccountType.dataset : AccountType.zoobc, payload: payload, display: a)
    }

    private static func hinted(_ a: String, _ hint: String) throws -> ParsedAddress {
        switch hint {
        case "eth":
            var h = a; if h.hasPrefix("0x") || h.hasPrefix("0X") { h = String(h.dropFirst(2)) }
            guard Enc.isHex(h, 40), let b = Enc.unhex(h) else { throw Invalid(message: "not a 20-byte Ethereum address") }
            return ParsedAddress(type: AccountType.ethereum, payload: b, display: a)
        case "sol":
            guard let d = Enc.base58Decode(a), d.count == 32 else { throw Invalid(message: "not a 32-byte Solana address") }
            return ParsedAddress(type: AccountType.solana, payload: d, display: a)
        case "dot":
            guard let ss = Enc.ss58Decode(a) else { throw Invalid(message: "not a valid SS58 address") }
            return ParsedAddress(type: AccountType.polkadot, payload: ss.1, display: a)
        case "zbc", "zbs": return try zbcForm(a)
        default: return try auto(a)
        }
    }

    private static func auto(_ a: String) throws -> ParsedAddress {
        let u = Array(a.utf8)
        if u.count == 42 && (a.hasPrefix("0x") || a.hasPrefix("0X")) && Enc.isHex(String(a.dropFirst(2)), 40) {
            return ParsedAddress(type: AccountType.ethereum, payload: Enc.unhex(String(a.dropFirst(2)))!, display: a)
        }
        if looksZbc(a) { return try zbcForm(a) }
        let low5 = String(a.prefix(5)).lowercased()
        if low5.hasPrefix("bc1") || low5.hasPrefix("tb1") || low5.hasPrefix("bcrt1") {
            guard let sw = Enc.segwitDecode(a) else { throw Invalid(message: "invalid Bitcoin bech32 address") }
            let (version, prog) = (sw.1, sw.2)
            switch (version, prog.count) {
            case (0, 20): return ParsedAddress(type: AccountType.bitcoinP2WPKH, payload: prog, display: a)
            case (0, 32): return ParsedAddress(type: AccountType.bitcoinP2WSH, payload: prog, display: a)
            case (1, 32): return ParsedAddress(type: AccountType.bitcoinTaproot, payload: prog, display: a)
            default: throw Invalid(message: "unsupported Bitcoin witness program")
            }
        }
        if (u[0] == UInt8(ascii: "1") || u[0] == UInt8(ascii: "3")) && u.count >= 26 && u.count <= 35 {
            if let raw = Enc.base58Decode(a), raw.count == 25, let body = Enc.base58CheckDecode(a), body.count == 21 {
                if body[0] == 0x00 { return ParsedAddress(type: AccountType.bitcoinP2PKH, payload: Array(body[1...]), display: a) }
                if body[0] == 0x05 { return ParsedAddress(type: AccountType.bitcoinP2SH, payload: Array(body[1...]), display: a) }
            }
        }
        if u.count > 5 && String(a.prefix(5)).lowercased() == "addr1" {
            if let d = Enc.bech32DecodePlain(a), d.0 == "addr", d.1.count == 29, d.1[0] == 0x61 { return ParsedAddress(type: AccountType.cardano, payload: Array(d.1[1...]), display: a) }
            throw Invalid(message: "invalid Cardano address (expected a mainnet enterprise addr1… address)")
        }
        if u[0] == UInt8(ascii: "T") && u.count == 34 {
            if let body = Enc.base58CheckDecode(a), body.count == 21, body[0] == 0x41 { return ParsedAddress(type: AccountType.tron, payload: Array(body[1...]), display: a) }
            throw Invalid(message: "invalid Tron address")
        }
        if u[0] == UInt8(ascii: "r") && u.count >= 25 && u.count <= 35 {
            if let body = Enc.base58CheckDecode(a, Enc.rippleAlphabet), body.count == 21, body[0] == 0x00 { return ParsedAddress(type: AccountType.ripple, payload: Array(body[1...]), display: a) }
            throw Invalid(message: "invalid Ripple address")
        }
        if a.hasPrefix("tz1") {
            if let body = Enc.base58CheckDecode(a), body.count == 23, body[0] == 0x06, body[1] == 0xa1, body[2] == 0x9f { return ParsedAddress(type: AccountType.tezos, payload: Array(body[3...]), display: a) }
            throw Invalid(message: "invalid Tezos address")
        }
        if let ss = Enc.ss58Decode(a), ss.1.count == 32 { return ParsedAddress(type: AccountType.polkadot, payload: ss.1, display: a) }
        if u.count >= 32 && u.count <= 44, let d = Enc.base58Decode(a), d.count == 32 { return ParsedAddress(type: AccountType.solana, payload: d, display: a) }
        if Enc.isHex(a, 64), let key = Enc.unhex(a) { return ParsedAddress(type: AccountType.zoobc, payload: key, display: encode(key, prefix: "ZBC")!) }
        throw Invalid(message: "unrecognised address. Supported: ZooBC (ZBC_/ZBS_), Bitcoin, Ethereum, Solana, Polkadot, Cardano, Ripple, Tron, Tezos")
    }

    /// Read a recipient in the order of spec/addresses.md section 3; `chain` forces one reading (--chain).
    public static func parse(_ text: String, chain: String = "") throws -> ParsedAddress {
        let a = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if a.isEmpty { throw Invalid(message: "empty address") }
        if chain.isEmpty { return try auto(a) }
        guard let hint = chains[chain.lowercased()] else { throw Invalid(message: "unknown chain \(chain)") }
        do { return try hinted(a, hint) } catch let e as Invalid {
            let up = a.uppercased(), low = a.lowercased()
            let plain = (a.utf8.count == 42 && (a.hasPrefix("0x") || a.hasPrefix("0X"))) || up.hasPrefix("ZBC") || up.hasPrefix("ZNK") || up.hasPrefix("ZBS")
                || a.hasPrefix("1") || a.hasPrefix("3") || low.hasPrefix("bc1") || low.hasPrefix("tb1") || low.hasPrefix("bcrt1") || a.utf8.count == 64
            if !plain { throw e }
            return try auto(a)
        }
    }

    /// A registry key parameter: 64 hex (optionally 0x) or a ZNK_/ZBG_/ZBR_/ZBC_ text address; 32 bytes.
    public static func parseKey32(_ text: String) throws -> [UInt8] {
        let u = Array(text.utf8)
        if u.count == 66 && u[3] == UInt8(ascii: "_") {
            guard let d = decode(text) else { throw Invalid(message: "invalid address checksum") }
            return d.1
        }
        var h = text; if h.hasPrefix("0x") || h.hasPrefix("0X") { h = String(h.dropFirst(2)) }
        guard Enc.isHex(h, 64), let b = Enc.unhex(h) else { throw Invalid(message: "key must be a 64-hex string or a ZNK_/ZBG_/ZBR_ address") }
        return b
    }
}
